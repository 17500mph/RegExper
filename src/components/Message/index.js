import React from 'react';
import PropTypes from 'prop-types';

import style from './style.module.css';

import InfoIcon from 'react-feather/dist/icons/info';
import ErrorIcon from 'react-feather/dist/icons/alert-octagon';
import WarningIcon from 'react-feather/dist/icons/alert-triangle';

const iconTypes = {
  info: InfoIcon,
  error: ErrorIcon,
  warning: WarningIcon
};

const renderIcon = (type, icon) => {
  const Icon = icon || iconTypes[type];

  if (!Icon) {
    return;
  }

  return <Icon />;
};

const Message = ({ type, icon, heading, children }) => (
  <div className={ [
    style.message,
    type && style[type]
  ].filter(Boolean).join(' ') }>
    <div className={ style.header }>
      { renderIcon(type, icon) }
      <h2>{ heading }</h2>
    </div>
    <div className={ style.content }>
      { children }
    </div>
  </div>
);

Message.propTypes = {
  type: PropTypes.oneOf([
    'info',
    'error',
    'warning'
  ]),
  icon: PropTypes.oneOfType([
    PropTypes.element,
    PropTypes.func
  ]),
  heading: PropTypes.string.isRequired,
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node
  ]).isRequired
};

export default Message;
