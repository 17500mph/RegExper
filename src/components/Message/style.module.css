@import url('../../globals.module.css');

.message {
  background: var(--color-tan);
  color: var(--color-black);
  margin: var(--spacing-margin) 0;
  box-shadow: 0 0 1rem color(var(--color-black) alpha(0.7));
  max-height: 100%;
  display: flex;
  flex-direction: column;
}

.header {
  & h2 {
    display: inline-block;
    padding: var(--spacing-margin);
    line-height: 2.8rem;
    margin: 0;
  }

  & svg {
    padding: var(--spacing-margin);
    height: 2.8rem;
    width: 2.8rem;
    vertical-align: bottom;
  }

  & button {
    padding: 0;
    margin: 0;
    border: 0 none;
    background: transparent;
    color: var(--color-tan);
    cursor: pointer;
    float: right;
    font-size: 0;

    &:hover,
    &:active {
      color: var(--color-white);
    }
  }
}

.content {
  padding: var(--spacing-margin);
  overflow: auto;

  & p {
    margin-top: 0;
    margin-bottom: var(--spacing-margin);
  }

  & p:last-child {
    margin-bottom: 0;
  }
}

.error {
  & .header {
    background: var(--color-red);
    color: var(--color-white);

    & svg {
      background: var(--color-white);
      color: var(--color-red);
    }
  }
}

.warning {
  & .header {
    background: var(--color-orange);

    & > svg {
      background: var(--color-black);
      color: var(--color-orange);
    }
  }
}

.info {
  & .header {
    background: var(--color-blue);
    color: var(--color-white);

    & > svg {
      background: var(--color-white);
      color: var(--color-blue);
    }
  }
}
