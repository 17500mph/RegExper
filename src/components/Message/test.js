import React from 'react';
import { shallow } from 'enzyme';

import Message from 'components/Message';

describe('Message', () => {
  test('rendering', () => {
    const component = shallow(
      <Message heading="Testing">
        <p>Message content</p>
      </Message>
    );
    expect(component).toMatchSnapshot();
  });

  test('rendering with icon', () => {
    const Icon = () => 'Sample icon SVG';
    const component = shallow(
      <Message heading="Testing" icon={ Icon }>
        <p>Message content</p>
      </Message>
    );
    expect(component).toMatchSnapshot();
  });

  test('rendering with type', () => {
    const component = shallow(
      <Message heading="Testing" type="error">
        <p>Message content</p>
      </Message>
    );
    expect(component).toMatchSnapshot();
  });
});
