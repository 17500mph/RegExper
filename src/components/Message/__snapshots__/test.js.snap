// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Message rendering 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Message
    heading="Testing"
  >
    <p>
      Message content
    </p>
  </Message>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="header"
        >
          <h2>
            Testing
          </h2>
        </div>,
        <div
          className="content"
        >
          <p>
            Message content
          </p>
        </div>,
      ],
      "className": "message",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            undefined,
            <h2>
              Testing
            </h2>,
          ],
          "className": "header",
        },
        "ref": null,
        "rendered": Array [
          undefined,
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Testing",
            },
            "ref": null,
            "rendered": "Testing",
            "type": "h2",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <p>
            Message content
          </p>,
          "className": "content",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Message content",
          },
          "ref": null,
          "rendered": "Message content",
          "type": "p",
        },
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="header"
          >
            <h2>
              Testing
            </h2>
          </div>,
          <div
            className="content"
          >
            <p>
              Message content
            </p>
          </div>,
        ],
        "className": "message",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              undefined,
              <h2>
                Testing
              </h2>,
            ],
            "className": "header",
          },
          "ref": null,
          "rendered": Array [
            undefined,
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Testing",
              },
              "ref": null,
              "rendered": "Testing",
              "type": "h2",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <p>
              Message content
            </p>,
            "className": "content",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Message content",
            },
            "ref": null,
            "rendered": "Message content",
            "type": "p",
          },
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`Message rendering with icon 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Message
    heading="Testing"
    icon={[Function]}
  >
    <p>
      Message content
    </p>
  </Message>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="header"
        >
          <Icon />
          <h2>
            Testing
          </h2>
        </div>,
        <div
          className="content"
        >
          <p>
            Message content
          </p>
        </div>,
      ],
      "className": "message",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <Icon />,
            <h2>
              Testing
            </h2>,
          ],
          "className": "header",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Testing",
            },
            "ref": null,
            "rendered": "Testing",
            "type": "h2",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <p>
            Message content
          </p>,
          "className": "content",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Message content",
          },
          "ref": null,
          "rendered": "Message content",
          "type": "p",
        },
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="header"
          >
            <Icon />
            <h2>
              Testing
            </h2>
          </div>,
          <div
            className="content"
          >
            <p>
              Message content
            </p>
          </div>,
        ],
        "className": "message",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <Icon />,
              <h2>
                Testing
              </h2>,
            ],
            "className": "header",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Testing",
              },
              "ref": null,
              "rendered": "Testing",
              "type": "h2",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <p>
              Message content
            </p>,
            "className": "content",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Message content",
            },
            "ref": null,
            "rendered": "Message content",
            "type": "p",
          },
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`Message rendering with type 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Message
    heading="Testing"
    type="error"
  >
    <p>
      Message content
    </p>
  </Message>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="header"
        >
          <AlertOctagon
            color="currentColor"
            size="24"
          />
          <h2>
            Testing
          </h2>
        </div>,
        <div
          className="content"
        >
          <p>
            Message content
          </p>
        </div>,
      ],
      "className": "message error",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <AlertOctagon
              color="currentColor"
              size="24"
            />,
            <h2>
              Testing
            </h2>,
          ],
          "className": "header",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "color": "currentColor",
              "size": "24",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Testing",
            },
            "ref": null,
            "rendered": "Testing",
            "type": "h2",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <p>
            Message content
          </p>,
          "className": "content",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Message content",
          },
          "ref": null,
          "rendered": "Message content",
          "type": "p",
        },
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="header"
          >
            <AlertOctagon
              color="currentColor"
              size="24"
            />
            <h2>
              Testing
            </h2>
          </div>,
          <div
            className="content"
          >
            <p>
              Message content
            </p>
          </div>,
        ],
        "className": "message error",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <AlertOctagon
                color="currentColor"
                size="24"
              />,
              <h2>
                Testing
              </h2>,
            ],
            "className": "header",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "color": "currentColor",
                "size": "24",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Testing",
              },
              "ref": null,
              "rendered": "Testing",
              "type": "h2",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <p>
              Message content
            </p>,
            "className": "content",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Message content",
            },
            "ref": null,
            "rendered": "Message content",
            "type": "p",
          },
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
